src/unix/ftrace.c

```diff
@@ -459,47 +575,200 @@ function_print_entry(struct ftrace_printer * p, struct rbuf_entry * entry)
+    return;
+}
+
+/*
+ * This must always called with rbuf already disabled
+ */
+__attribute__((no_instrument_function))
+static void
+function_graph_trace_entry(struct ftrace_graph_entry * stack_entry)
+{
+    struct rbuf_entry * entry;
+    struct rbuf_entry_function_graph * graph;
+
+    rbuf_lock(&global_rbuf);
+    if (!__rbuf_acquire_write_entry(&global_rbuf, &entry))
+        goto drop;
+
+    graph = &(entry->graph);
+    graph->ip = stack_entry->func;
+    graph->duration = 0;
+    graph->cpu = 0;
+    graph->depth = stack_entry->depth;
+    graph->has_child = 1;
+    runtime_memcpy(graph->name, current->name, 15);
+
+    rbuf_unlock(&global_rbuf);
+    return;
+drop:
+    /* XXX count the drop */
+    return;
+}
+
+/*
+ * This must always called with rbuf already disabled
+ */
+__attribute__((no_instrument_function))
+static void
+function_graph_trace_return(struct ftrace_graph_entry * stack_entry)
+{
+    struct rbuf_entry * entry;
+    struct rbuf_entry_function_graph * graph;
+
+    rbuf_lock(&global_rbuf);
+    if (!__rbuf_acquire_write_entry(&global_rbuf, &entry))
+        goto drop;
+
+    graph = &(entry->graph);
+    graph->depth = stack_entry->depth;
+    graph->ip = stack_entry->func;
+    graph->duration = (stack_entry->return_ts - stack_entry->entry_ts);
+    graph->cpu = 0;
+    graph->has_child = stack_entry->has_child;
+
+    /* only need the name if we have no children, otherwise
+      * we already have it in the entry call
+      */
+    if (!graph->has_child)
+        runtime_memcpy(graph->name, current->name, 15);
+
+    rbuf_unlock(&global_rbuf);
+    return;
+drop:
+    /* XXX count the drop */
+    return;
+}
+
+__attribute__((no_instrument_function))
+static void
+function_graph_toggle(boolean enable)
+{
+    if (enable) {
+        __ftrace_graph_entry_fn = function_graph_trace_entry;
+        __ftrace_graph_return_fn = function_graph_trace_return;
+    } else {
+        __ftrace_graph_entry_fn = (ftrace_graph_t)ftrace_stub;
+        __ftrace_graph_return_fn = (ftrace_graph_t)ftrace_stub;
+    }
+}
+
+static void
+print_switch_event(struct ftrace_printer * p, struct rbuf_entry_switch * sw)
+{
+    printer_write(p, "------------------------------------------\n");
+    printer_write(p, " %d) %s-%d  => %s-%d\n",
+        sw->cpu,
+        runtime_strlen(sw->name_out) ? sw->name_out : "tid",
+        sw->tid_out,
+        runtime_strlen(sw->name_in) ? sw->name_in : "tid",
+        sw->tid_in
+    );
+    printer_write(p, "------------------------------------------\n");
+}
+
+static void
+function_graph_print_entry(struct ftrace_printer * p,
+                           struct rbuf_entry * entry)
+{
+    struct rbuf_entry_function_graph * graph = &(entry->graph);
+
+    /* check for special switch event */
+    if (graph->depth == TRACE_GRAPH_SWITCH_DEPTH) {
+        print_switch_event(p, &(entry->sw));
+        return;
+    }
+
+    printer_write(p, " %d) ", graph->cpu);
+
+    /* XXX symbol */
```

Comment:
What is this in reference to?

Reply:
linux adds various symbols that correspond to latency ranges, making it easier to search for long running functions. I'll add it in, not a big effort